### ISSUE TITLE
Reactive mongo hangs with inheritance hierarchies defined in different gradle project

### ISSUE THREAD

**Issue by martin-drozdik** [spring-projects/spring-boot#12875]
I made a minimal working example for this issue at:

https://github.com/martin-drozdik/spring-mongo-bug

I am not completely sure whether this is a single bug, or two different manifestations of the same bug.

Instructions to reproduce:

1. Check out branch `master` of the given repository
2. Run an instance of mongodb 3.6 on the classical port
3. You will have a file called `Controller.kt` which looks roughly like this:

```
/*
@TypeAlias("ServerEvent")
sealed class ServerEvent

@Document
@TypeAlias("UserRegistered")
data class UserRegistered
(
    @Indexed val something: String

): ServerEvent()
*/
@Repository
interface ServerEventRepository: ReactiveCrudRepository<ServerEvent, String>

@RestController
@RequestMapping
class Controller(private val serverEventRepository: ServerEventRepository)
{
    init
    {
        // In order for the bug to manifest itself, there has to be at least one item
        // in the database.
        // Once you fill the database, you should comment out this init section and
        // restart the application.
        serverEventRepository.deleteAll().subscribe()
        serverEventRepository.save(UserRegistered("lala")).subscribe()
    }

    @GetMapping
    fun get(): Mono<Int>
    {
        // The bug only manifests itself with functions that construct the UserRegistered object.
        // Therefore I use this contrived example as opposed to 'serverEventRepository.count()'.

        return serverEventRepository.findAll().collectList().map { it.size }
    }
}
```
4. The `ServerEvent` is commented out and included in a different gradle sub-project in a multi-project setup.
5. Start the application and do a single  `curl -v localhost:8080`. This will cause one `UserRegistered` object to be inserted into mongo.
6. Comment out the `init` block to prevent any modification to the DB henceforth and restart the application.
6. `curl -v localhost:8080` 
7. The application gets stuck in the middle of a HTTP request with:

```
2018-04-16 18:06:57.935 DEBUG 26355 --- [ctor-http-nio-2] r.i.n.http.server.HttpServerOperations   : Increasing pending responses, now 1
2018-04-16 18:06:57.947 DEBUG 26355 --- [ctor-http-nio-2] r.ipc.netty.http.server.HttpServer       : [id: 0xfeb89506, L:/127.0.0.1:8080 - R:/127.0.0.1:48722] READ COMPLETE
2018-04-16 18:06:57.948 DEBUG 26355 --- [ctor-http-nio-2] r.ipc.netty.channel.ChannelOperations    : [HttpServer] [id: 0xfeb89506, L:/127.0.0.1:8080 - R:/127.0.0.1:48722] handler is being applied: org.springframework.http.server.reactive.ReactorHttpHandlerAdapter@1212ad72
2018-04-16 18:06:57.975 DEBUG 26355 --- [ctor-http-nio-2] o.s.web.reactive.DispatcherHandler       : Processing GET request for [http://localhost:8080/]
2018-04-16 18:06:57.996 DEBUG 26355 --- [ctor-http-nio-2] s.w.r.r.m.a.RequestMappingHandlerMapping : Looking up handler method for path /
2018-04-16 18:06:57.998 DEBUG 26355 --- [ctor-http-nio-2] s.w.r.r.m.a.RequestMappingHandlerMapping : Returning handler method [public reactor.core.publisher.Mono<java.lang.Integer> application.Controller.get()]
2018-04-16 18:06:57.998 DEBUG 26355 --- [ctor-http-nio-2] o.s.b.f.s.DefaultListableBeanFactory     : Returning cached instance of singleton bean 'controller'
2018-04-16 18:06:58.072 DEBUG 26355 --- [ctor-http-nio-2] o.s.d.m.core.ReactiveMongoTemplate       : find using query: { } fields: Document{{}} for class: class domain.ServerEvent in collection: serverEvent
2018-04-16 18:06:58.215  INFO 26355 --- [ntLoopGroup-2-2] org.mongodb.driver.connection            : Opened connection [connectionId{localValue:3, serverValue:190}] to localhost:27018
2018-04-16 18:06:58.224 DEBUG 26355 --- [ntLoopGroup-2-2] org.mongodb.driver.protocol.command      : Asynchronously sending command {find : BsonString{value='serverEvent'}} to database test on connection [connectionId{localValue:3, serverValue:190}] to server localhost:27018
2018-04-16 18:06:58.261 DEBUG 26355 --- [ntLoopGroup-2-2] o.s.beans.CachedIntrospectionResults     : Not strongly caching class [domain.UserRegistered] because it is not cache-safe
2018-04-16 18:06:58.278 DEBUG 26355 --- [ntLoopGroup-2-2] .m.c.i.MongoPersistentEntityIndexCreator : Analyzing class class domain.UserRegistered for index information.
2018-04-16 18:06:58.305 DEBUG 26355 --- [ntLoopGroup-2-2] org.mongodb.driver.protocol.command      : Asynchronously sending command {createIndexes : BsonString{value='userRegistered'}} to database test on connection [connectionId{localValue:3, serverValue:190}] to server localhost:27018

```

The problem manifests itself in `2.0.1.RELEASE` and `2.0.0.RELEASE`, but *not* in `2.0.0.M7`. 

If I move the class definitions of `UserRegistered` to the same gradle project (see `same-project` git branch), the problem goes away.

The problem manifests itself only with inheritance and can be workarounded by either removing the `@Indexed` annotation, removing the `@Document` annotation, or by adding a `@TypeAlias` annotation. In the latter case however I get a different error:

```
{"timestamp":"2018-04-16T15:01:59.002+0000","path":"/","status":500,"error":"Internal Server Error","message":"java.lang.IllegalAccessException: Class kotlin.reflect.jvm.internal.FunctionCaller$Constructor can not access a member of class domain.ServerEvent with modifiers \"private\""}
```

Which may not be related to this issue.

I will be happy to provide more details or to make more experiments.

**Comment by snicoll** [comment]
Thanks for the sample and the details but, given the workaround that you've indicated, this is the wrong place to report the issue. Can you please raise that against [the spring data mongodb project](https://jira.spring.io/browse/DATAMONGO/)?

**Comment by kjq** [comment]
Holy crap - I just spent days on this.  I never bothered to think that it was hanging on the index stuff (even though that is where it hung in my logs)  Thank you for the hint!

When I removed all the index annotations (@Index and @CompoundIndex), at least from the 2nd DAO my chained call worked.

Is there a ticket for this in JIRA?

This was my original problem - https://stackoverflow.com/questions/49930258/how-to-chain-2-spring-mongo-reactive-save-operations-together
